python/ray/worker.py

```diff
@@ -1968,30 +1982,72 @@ def _register_class(cls, use_pickle=False, worker=global_worker):
 
     Args:
         cls (type): The class that ray should serialize.
-        use_pickle (bool): If False then objects of this class will be
-            serialized by turning their __dict__ fields into a dictionary. If
-            True, then objects of this class will be serialized using pickle.
+        use_pickle (bool): If true, then objects of this class will be
+            serialized using pickle.
+        use_dict: If true, then objects of this class be serialized turning
+            their __dict__ fields into a dictionary. Must be False if
+            use_pickle is true.
+        serializer: The custom serializer to use. This should be provided if
+            and only if use_pickle and use_dict are False.
+        deserializer: The custom deserializer to use. This should be provided
+            if and only if use_pickle and use_dict are False.
+        local: True if the serializers should only be registered on the current
+            worker. This should usually be False.
 
     Raises:
         Exception: An exception is raised if pickle=False and the class cannot
-            be efficiently serialized by Ray.
+            be efficiently serialized by Ray. This can also raise an exception
+            if use_dict is true and cls is not pickleable.
     """
-    if not use_pickle:
+    assert not (use_pickle and use_dict), ("If use_pickle is true, then "
```

Comment:
consider doing

custom_serializer = serializer != None and deserializer != None

assert custom_serializer + use_pickle + use_dict == 1, "Exactly one of ... needs to be specified"